Add QMC5883L Sensor + Improvements to HMC5883L

timpur opened the pull request:
## Description:


**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).


Added QMC5883L sensor and improvements to the HMC5883L sensor

Docs: https://github.com/esphome/esphome-docs/pull/301

Commit:
Add QMC5883L and Updated HMC5883L

Commit:
add tests

Commit:
changed to oversampling

Commit:
fix pylint

Commit:
fix private method

Commit:
typo fix

Commit:
fix protected method

timpur commented:
getting this issue in build checks, not my code ??
```
-     msg.effects.push_back("None");
+    msg.effects.emplace_back("None");
```

aneisch commented:
Been testing/using the QMC5883L sensor and it works flawlessly

OttoWinter commented on esphome/components/hmc5883l/sensor.py:
```diff
@@ -42,11 +72,16 @@ def validate_range(value):
+        HMC5883LDatarates, unit="Hz", int=False),
+    cv.Optional(CONF_MEASUREMENT_MODE, default='normal'): validate_enum(
+        HMC5883LMeasurementModes, int=False),
+    cv.Optional(CONF_RANGE, default='130uT'): validate_enum(HMC5883L_RANGES, unit="uT"),
```
This should accept both µT and uT (otherwise this would be a breaking change).

OttoWinter commented on esphome/components/qmc5883l/qmc5883l.cpp:
```diff
@@ -0,0 +1,140 @@
+    return;
+  }
+
+  // if (id != 0x01) {
```
Please remove commented-out code or add comments for future use.

OttoWinter commented on esphome/components/qmc5883l/qmc5883l.cpp:
```diff
@@ -0,0 +1,140 @@
+}
+
+bool QMC5883LComponent::read_byte_16_(uint8_t a_register, uint16_t *data, uint32_t conversion) {
+  bool success = QMC5883LComponent::I2CDevice::read_byte_16(a_register, data, conversion);
```
```suggestion
  bool success = this->read_byte_16(a_register, data, conversion);
```

OttoWinter commented on esphome/components/qmc5883l/sensor.py:
```diff
@@ -0,0 +1,87 @@
+def validate_enum(enum_values, unit=None, int=True):
+    def validate_enum_bound(value):
+        value = cv.string(value)
+        if unit and (value.endswith(unit.encode(encoding='UTF-8', errors='strict'))
```
`unit.encode` should not be necessary here - `value` is already a unicode string (py2: unicode, py3: str)

timpur commented:
Thanks @OttoWinter, will try get this done this weekend.

Commit:
Clean up code and PR recomendations

timpur commented:
@OttoWinter Believe this is ready fir review again thx :)

Commit:
fix tests

Commit:
remote file

Commit:
merge

Commit:
fix qmc oversampling unit

OttoWinter committed:
Remove hmc5883l config logging

Either the units are converted to the user values like 1x, 8x oversampling or not printed at all. Printing the machine-value of these is only confusing users.

OttoWinter committed:
Changes for validate_enum

Move stuff that can be done beforehand out of the bound function, use text_type for py2/3 compatability.

OttoWinter committed:
Remove unused constant

OttoWinter committed:
Remove duplicate tests

OttoWinter committed:
Merge branch 'qmc5883l' of https://github.com/timpur/esphome into qmc5883l

OttoWinter committed:
Repeat remove config print

Commit:
remove changes to test2 since bin is to large

Commit:
Merge branch 'qmc5883l' of https://github.com/timpur/esphome into qmc5883l

timpur commented on esphome/components/hmc5883l/sensor.py:
```diff
@@ -29,11 +50,21 @@
-    if value.endswith(u'µT') or value.endswith('uT'):
-        value = value[:-2]
-    return cv.enum(HMC5883L_RANGES, int=True)(value)
+def validate_enum(enum_values, units=None, int=True):
```
@OttoWinter at some point should utils like this, be made common. Think there could be a bit of clean up in the code base for sensors .... not sure what your thoughts are on this?

OttoWinter replied:
Yeah I guess so, `config_validation.py` would be the place for it (maybe call it `enum_with_unit`?)

timpur committed:
Add comment to HMC5583L

timpur commented:
@OttoWinter Did you want to clean up now or later ?

OttoWinter commented:
@timpur What do you mean?

pilotak commented on esphome/components/qmc5883l/qmc5883l.cpp:
```diff
@@ -0,0 +1,124 @@
+  const float y = int16_t(raw_y) * mg_per_bit * 0.1f;
+  const float z = int16_t(raw_z) * mg_per_bit * 0.1f;
+
+  float heading = atan2f(0.0f - x, y) * 180.0f / M_PI;
```
it would be nice to take in place a position you have a board mounted https://github.com/gcmike/QMC5883/blob/master/QMC5883.cpp#L89-L100

timpur replied:
You cold always have a custom sensor that uses which ever axis you like in your yaml to create the desired heading orientation.

pilotak replied:
Fair enough

pilotak commented:
I have checked QMC on ESP8266 and also on ESP32, it works like a charm.
